### ISSUE TITLE
Async AccessEvent logging to OutputStreamAppenders has some issues with Jetty

### ISSUE THREAD

**Issue by alex-sherwin** [logstash/logstash-logback-encoder#94]
The problem occurs because the Jetty web server "recycles" `org.eclipse.jetty.server.Request` Objects, when the server is done with a given `Request` Object it calls `org.eclipse.jetty.server.Request.recycle()`

When using `AccessEventAsyncDisruptorAppender`, the async behavior creates a race condition when trying to read anything that the `Request.recycle()` method has cleared out, for example a `HttpServletRequest` attribute configured in a JsonEncoder as `%requestAttribute{MY_PARAM}`

I haven't found a way to configure Jetty to modify it's behavior to make this workable for myself while still using an async appender, so there might be no solution here.

At the very least the documentation should outline this

**Comment by alex-sherwin** [comment]
Upon further investigation it actually seems to be caused by the way that the `AccessEventAsyncDisruptorAppender` ends up calling `AccessEvent.prepareForDeferredProcessing()` twice.  

It is first called on the web request thread, which ends up calling `AccessEvent.copyAttributeMap()` and copies all the `HttpServletRequest` attributes into a local map.

After it's handed off to the `AccessEventAsyncDisruptorAppender`, the thread that runs which is actually responsible for writing the log invokes `AccessEvent.prepareForDeferredProcessing()` again, which calls `AccessEvent.copyAttributeMap()` again, which blindly wipes out the properly copied local attribute map copy, and attempts to re-populate it with the references jetty request (which has now been recycled) and may or may not get some or none of the values copied (it's a race with the jetty recycle cleanup at this point)

I was able to work around the issue by creating a custom `RequestLogImpl` subclass that overrides the `log(..)` method, and it applies the same logic as before but calls `new AccessEvent(..)`  with a custom wrapped `HttpServletRequestWrapper` which has it's own local attribute map copy.

Using this technique, now when the async thread calls `AccessEvent.copyAttributeMap()` the 2nd time, it will be operating on the custom `HttpServletRequestMapper` which has a saved copy of the attribute map

**Comment by philsttr** [comment]
`AccessEventAsyncDisruptorAppender` must call `accessEvent.prepareForDeferredProcessing()` in the request thread because the event is going to be handed off to another thread.

However, I don't see where `AccessEventAsyncDisruptorAppender` is ever calling `accessEvent.prepareForDeferredProcessing()` in the worker thread.  The only thing the worker thread does with the event is call the delegate appenders (see [here](https://github.com/logstash/logstash-logback-encoder/blob/master/src/main/java/net/logstash/logback/appender/DelegatingAsyncDisruptorAppender.java#L53)).

To what appender are you delegating?  Can you post your logback config?

As a side note, it looks like this problem was introduced in logback 1.1.3 via [LOGBACK-1033](http://jira.qos.ch/browse/LOGBACK-1033) in pull request [240](https://github.com/qos-ch/logback/pull/240).  Depending on what appender you're delegating to, the proper fix might need to be in logback itself, not in logstash-logback-encoder.

**Comment by alex-sherwin** [comment]
Here's the `logback-access.xml` that exhibited the issue for me:

    <?xml version="1.0" encoding="UTF-8" standalone="yes"?>
    <configuration debug="false" scan="false">
    
      <statusListener class="ch.qos.logback.core.status.NopStatusListener"/>
      
    
      <appender name="logstash-access" class="net.logstash.logback.appender.AccessEventAsyncDisruptorAppender">
    
        <appender class="ch.qos.logback.core.rolling.RollingFileAppender">
    
          <file>/somefile.log</file>
          <rollingPolicy class="ch.qos.logback.core.rolling.TimeBasedRollingPolicy">
            <fileNamePattern>/archive/somefile.log.%d</fileNamePattern>
            <maxHistory>60</maxHistory>
          </rollingPolicy>
    
          <encoder class="net.logstash.logback.encoder.AccessEventCompositeJsonEncoder">
    
            <providers>
    
              <timestamp>
                <pattern>yyyy-MM-dd'T'HH:mm:ss.SSSZZ</pattern>
                <timeZone>UTC</timeZone>
              </timestamp>
    
              <pattern>
                <pattern>
                  {
                  "app": "someapp",
                  "request_custom1": "%requestAttribute{myprefix.custom1}",
                  "request_custom2": "%requestAttribute{myprefix.custom2}",
                  "request_custom3": "%requestAttribute{myprefix.custom3}",
                  "request_custom4": "%requestAttribute{myprefix.custom4}",
                  "node": "${SOME_ENV_VAR}",
                  "type": "access",
                  "remote_ip": "%a",
                  "local_ip": "%A",
                  "local_port": "#asLong{%localPort}",
                  "status_code": "#asLong{%s}",
                  "protocol": "%H",
                  "method": "%m",
                  "request_uri": "#nullNA{%U}",
                  "request_url": "#nullNA{%r}",
                  "elapsed_time": "#asLong{%D}",
                  "user_agent": "#nullNA{%header{User-Agent}}",
                  "accept": "#nullNA{%header{Accept}}",
                  "referer": "#nullNA{%header{Referer}}",
                  "bytes_sent": "#asLong{%b}"
                  }
                </pattern>
              </pattern>
    
    
            </providers>
    
          </encoder>
    
        </appender>
    
      </appender>
    
      <appender-ref ref="logstash-access"/>
    
    </configuration>


It was against Jetty `9.2.10.v20150310`, logback/logback-access `1.1.3` and logstash-logback-encoder `4.3`

I was able to consistently observe on every request that `accessEvent.prepareForDeferredProcessing()` was being invoked twice (once by the request thread and again later by the thread inside the async writer thread)

**Comment by philsttr** [comment]
I have reproduced this myself.

The problem occurs because logback's `RollingFileAppender` calls `prepareForDeferredProcessing` [here](https://github.com/qos-ch/logback/blob/master/logback-core/src/main/java/ch/qos/logback/core/OutputStreamAppender.java#L212)

Here is the stacktrace where it occurs:

```
Daemon Thread [logback-async-disruptor-appender-1] (Suspended (breakpoint at line 533 in AccessEvent))	
	AccessEvent.prepareForDeferredProcessing() line: 533	
	RollingFileAppender<E>(OutputStreamAppender<E>).subAppend(E) line: 212	
	RollingFileAppender<E>.subAppend(E) line: 182	
	RollingFileAppender<E>(OutputStreamAppender<E>).append(E) line: 103	
	RollingFileAppender<E>(UnsynchronizedAppenderBase<E>).doAppend(E) line: 88	
	AppenderAttachableImpl<E>.appendLoopOnAppenders(E) line: 48	
	DelegatingAsyncDisruptorAppender$DelegatingEventHandler.onEvent(LogEvent<Event>, long, boolean) line: 53	
	DelegatingAsyncDisruptorAppender$DelegatingEventHandler.onEvent(Object, long, boolean) line: 49	
	BatchEventProcessor<T>.run() line: 128	
	Executors$RunnableAdapter<T>.call() line: 471	
	ScheduledThreadPoolExecutor$ScheduledFutureTask<V>(FutureTask<V>).run() line: 262	
	ScheduledThreadPoolExecutor$ScheduledFutureTask<V>.access$201(ScheduledThreadPoolExecutor$ScheduledFutureTask) line: 178	
	ScheduledThreadPoolExecutor$ScheduledFutureTask<V>.run() line: 292	
	ScheduledThreadPoolExecutor(ThreadPoolExecutor).runWorker(ThreadPoolExecutor$Worker) line: 1145	
	ThreadPoolExecutor$Worker.run() line: 615	
	Thread.run() line: 745	

```

Unfortunately, since this is in logback's code, I don't think there's anything that can be done in logstash-logback-encoder to fix this.

The fix for this will need to be in logback itself.  Please submit a bug against logback.

Until this is fixed in logback, you'll need to either
* NOT use asynchronous logging
* continue with your workaround

**Comment by dimas** [comment]
I would not blame my PR https://github.com/qos-ch/logback/pull/240 - it actually allows request attributes to be logged properly. Without it it would not work even with TCP appender. 

Obviously the problem is the fact that prepareForDeferredProcessing is not re-enterant and does not understand the job is already done. So the proper fix would be to modify AccessEvent.java to remember the fact prepareForDeferredProcessing() was called already and does nothing on the second call. Or at least raising an issue in logback issue tracker.

**Comment by philsttr** [comment]
Thanks @dimas.  Yeah, you're right.  That PR didn't _cause_ the problem.  It just exposed the bug in logback.

I'm going to close this issue here, since the proper fix is in logback.

@alex-sherwin Feel free to open an issue / submit a PR against logback.

**Comment by ceki** [comment]
The fix will be available with logback version 1.1.10.
